gnmidiff/setrequest.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// gnmidiff contains gNMI utilities for diffing SetRequests and GetResponses.
+package gnmidiff
+
+import (
+	"fmt"
+
+	"github.com/derekparker/trie"
+	"github.com/openconfig/ygot/ygot"
+	"github.com/openconfig/ygot/ytypes"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// MismatchedUpdate represents two different update values for the same leaf
+// node.
+type MismatchedUpdate struct {
+	// A is the update value in A.
+	A interface{}
+	// B is the update value in B.
+	B interface{}
+}
+
+// SetRequestDiff contains the intent difference between two SetRequests.
+//
+// - The key of the maps is the string representation of a gpb.Path constructed
+// by ygot.PathToString.
+// - The value of the update fields is the JSON_IETF representation of the
+// value.
+type SetRequestDiff struct {
+	AOnlyDeletes      map[string]struct{}
+	BOnlyDeletes      map[string]struct{}
+	CommonDeletes     map[string]struct{}
+	AOnlyUpdates      map[string]interface{}
+	BOnlyUpdates      map[string]interface{}
+	CommonUpdates     map[string]interface{}
+	MismatchedUpdates map[string]MismatchedUpdate
+}
+
+// DiffSetRequest returns a unique and minimal intent diff of two SetRequests.
+//
+// newSchema is intended to be provided via the function defined in generated
+// ygot code (e.g. exampleoc.Schema).
+// If newSchema is nil, then DiffSetRequest will make the following assumption:
+// - Any JSON value in the input SetRequest MUST conform to the OpenConfig
+// YANG style guidelines. See the following for checking compliance.
+// * https://github.com/openconfig/oc-pyang
+// * https://github.com/openconfig/public/blob/master/doc/openconfig_style_guide.md
+//
+// Currently, support is only for SetRequests without any delete paths, and
+// replace and updates that don't have conflicting leaf values. If not
+// supported, then an error will be returned.
+func DiffSetRequest(a *gpb.SetRequest, b *gpb.SetRequest, newSchema func() (*ytypes.Schema, error)) (SetRequestDiff, error) {
```

Comment:
The title here says non-schema aware, but we take in the schema. Maybe I'm missing something here but can you explain the reason `newSchema` is needed here?

Reply:
I have `// TODO: implement newSchema handling.`
I will get this done in a later PR.

Reply:
implementation at https://github.com/openconfig/ygot/pull/776